eccodes: add v2.32.0, v2.31.0

victoria-cherkas opened the pull request:
Add eccodes 2.32.0 and 2.31.0

Comment re aec change to default true: https://github.com/ecmwf/eccodes/blob/develop/CMakeLists.txt#L133-L143
    # AEC (Adaptive Entropy Coding) provides the WMO GRIB CCSDS compression and decompression of data.
    # This is highly recommended from ecCodes >= 2.25.0
    # To force the build without it, use -DENABLE_AEC=OFF

climbfuji committed:
Fix version incompatibilities of py-pandas and py-openpyxl (#40472)

* Fix version incompatibilities of py-pandas and py-openpyxl

* Add variant excel for py-pandas

* Add package py-pyxlsb

* Add versios for py-xlsxwriter

* Define excel dependencies for py-pandas 1.4, 1.5, 2.0, 2.1

* Fix variant excel in py-pandas

* Add package py-odfpy, which is also a dependency for py-pandas@2.0:

* Rearrange excel dependencies for py-pandas

* Change url to pypi

* Add missing newline to fix style in py-odfpy

victoria-cherkas committed:
add latest eccodes versions

renjithravindrankannath committed:
Updating patch to enable flag mcode-object-version=none (#40367)

* Updating patch to add flag mcode-object-version=none when
   device libs is buils as part of llvm-amdgpu
* Limiting patch to +rocm-device-libs variant and adding
   appropriate comment for the patch
* Updating llvmpatch as per the mailine code
   Updating hsa-rocr patch as per the latest code
   Updating the if elif condition for the hip test src path
* Updating flags for 5.5 relases and above
* Updating build flags and patches

tldahlgren commented on var/spack/repos/builtin/packages/eccodes/package.py:
```diff
@@ -317,7 +320,7 @@ def cmake_args(self):
             self.define("ENABLE_JPG_LIBOPENJPEG", jp2k == "openjpeg"),
             self.define_from_variant("ENABLE_PNG", "png"),
-            self.define_from_variant("ENABLE_AEC", "aec"),
+            self.define("ENABLE_AEC", "aec" if "+aec" in self.spec else "OFF"),
```
Why did this line change (`aec` is a variant)?

victoria-cherkas replied:
Followed the pattern of BUILD_SHARED_LIBS:
`self.define("BUILD_SHARED_LIBS", "BOTH" if "+shared" in self.spec else "OFF"),`
because to force the build without aec we now need to define -DENABLE_AEC=OFF

I wasn't sure whether `define_from_variant` where variant is boolean, would result in an 'ON'/'OFF' value for `ENABLE_AEC`?

victoria-cherkas replied:
Otherwise thanks for the review!

haampie replied:
It does: `+aec` -> `ON`, `~aec` -> `OFF`. It's in fact "better", since it adds the type: `-DENABLE_AEC:BOOL=...` automatically.

victoria-cherkas replied:
Ok, thanks a lot for the explanation. Returned to `define_from_variant` in https://github.com/spack/spack/pull/40564/commits/c41159b46734e512267c7c0b5aa90b7edba193aa

tldahlgren reviewed (commented):
Confirmed the two new version sha256.  Do question change from `define_from_variant()` to `define()`.  Also deferring to @skosukhin since this PR changes the default build options.

adamjstewart committed:
py-lightning: add v2.1.0 (#40496)

adamjstewart committed:
py-grpcio: cython 3 still not supported (#40537)

adamjstewart committed:
py-shapely: add v2.0.2 (#40523)

adamjstewart committed:
py-fiona: add v1.9.5 (#40497)

alalazo committed:
Use string representation of deptypes for concrete specs (#40566)

adamjstewart committed:
py-grayskull: add new package (#40293)

* py-grayskull: add new package

* [@spackbot] updating style on behalf of adamjstewart

---------

Co-authored-by: adamjstewart <adamjstewart@users.noreply.github.com>

sethrj committed:
py-furo: new version (#40559)

Commit:
intel-oneapi-compilers: ifx is located in bin not bin/intel64 (#40561)

This is a fix on top of https://github.com/spack/spack/pull/40557 .
Tagging @rscohn2 for review.

berquist committed:
Add latest versions of rlwrap (#40563)

* Add latest versions of rlwrap
* rlwrap: fix URL for v0.46.1

patrickb314 committed:
Creation of Beatnik package and associated updates to silo and cabana spack package (#40382)

* Added initial package for building Beatnik with spack

* Fixed github ID for Jason as a maintainer.

* Major revision of beatnik spack package to properly support GPU spack builds with CUDA (and ROCm, though that it untested)

* Marked that beatnik 1.0 will require cabana 0.6.0. We will wait for the cabana 0.6.0 release before we release beatnik

* Update to beatnik package spec to compile with hipcc when +rocm

* Updated spack package for cabana for version 0.6.0 and appropriate heffte dependency

* Updated beatnik package to require cabana 0.6.0

* More updates to cabana and beatnik to build with cabana 0.6.0

* Finish removing BLT dependence from beatnik

* More updates to beatnik package for compiling on cray systems

* Updated beatnik package for new cabana package

* Changes to silo package for new silo version

* Fixed version specs for heffte to be able to concretize and build

* Fixed spack style issues for beatnik and silo packages

* More spack formatting fixes to beatnik and silo

* Patrick adopting silo package as maintainer for now

* Should address final style changes to beatnik package spec

* Yet more style fixes.

* Perhaps this is the final style fixes? :)

* Minor fix to cabana package on required versions

alalazo committed:
packages: use "requires" to allow only selected compilers (#40567)

A few packages have encoded an idiom that pre-dates the introduction
of the 'requires' directive, and they cycle over all compilers
to conflict with the ones that are not supported.

Here instead we reverse the logic, and require the ones that
are supported.

victoria-cherkas committed:
revert to define_from_variant for ENABLE_AEC

haampie committed:
Support `spack env activate --with-view <name> <env>` (#40549)

Currently `spack env activate --with-view` exists, but is a no-op.

So, it is not too much of a breaking change to make this redundant flag
accept a value `spack env activate --with-view <name>` which activates
a particular view by name.

The view name is stored in `SPACK_ENV_VIEW`.

This also fixes an issue where deactivating a view that was activated
with `--without-view` possibly removes entries from PATH, since now we
keep track of whether the default view was "enabled" or not.

wspear committed:
Support apple-clang in pdt (#40582)

haampie committed:
patchelf: fix compilation with GCC 7 (#40581)

Commit:
hiop: 1.0.1 release (#40580)

adamjstewart committed:
py-rtree: add v1.1.0 (#40575)

RMeli committed:
Add mpi_f08 variant to CP2K (#40574)

* add mpi_f08 variant
* add conflict
* add conflict with released versions of cp2k and +mpi_f08

scheibelp committed:
Allow / in GitVersion (#39398)

This commit allows version specifiers to refer to git branches that contain
forward slashes. For example, the following is valid syntax now:

    pkg@git.releases/1.0
   
It also adds a new method `Spec.format_path(fmt)` which is like `Spec.format`,
but also maps unsafe characters to `_` after interpolation. The difference is
as follows:

    >>> Spec("pkg@git.releases/1.0").format("{name}/{version}")
    'pkg/git.releases/1.0'

    >>> Spec("pkg@git.releases/1.0").format_path("{name}/{version}")
    'pkg/git.releases_1.0'

The `format_path` method is used in all projections. Notice that this method
also maps `=` to `_`

    >>> Spec("pkg@git.main=1.0").format_path("{name}/{version}")
    'pkg/git.main_1.0'
   
which should avoid syntax issues when `Spec.prefix` is literally copied into a
Makefile as sometimes happens in AutotoolsPackage or MakefilePackage

msimberg committed:
Add tracy 0.10 (#40573)

dennisklein committed:
xmlto: add more dependencies (#40578)

`xmllint` is called by `xmlto` during generation of `libzmq`'s docs, so
adding `libxml2`.

The docbook deps and the patches are taken from
https://src.fedoraproject.org/rpms/xmlto/blob/rawhide/f/xmlto.spec

There are still many more dependencies missing, but this is out of scope
of this patch (which is only concerned about the use case of `libzmq`).

jfavre committed:
paraview: add variant for NVIDIA IndeX (#40577)

* add variant for NVIDIA IndeX
* remove whitespaces

becker33 committed:
abi.py: fix typo, add type-hints (#38216)


Co-authored-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

boomanaiden154 committed:
Add license directive (#39346)

This patch adds in a license directive to get the ball rolling on adding in license 
information about packages to spack. I'm primarily interested in just adding
license into spack, but this would also help with other efforts that people are
interested in such as adding license information to the ASP solve for 
concretization to make sure licenses are compatible.

Usage:

Specifying the specific license that a package is released under in a project's
`package.py` is good practice. To specify a license, find the SPDX identifier for
a project and then add it using the license directive:

```python
   license("<SPDX Identifier HERE>")
```

For example, for Apache 2.0, you might write:

```python
   license("Apache-2.0")
```

Note that specifying a license without a when clause makes it apply to all
versions and variants of the package, which might not actually be the case.
For example, a project might have switched licenses at some point or have
certain build configurations that include files that are licensed differently.
To account for this, you can specify when licenses should be applied. For
example, to specify that a specific license identifier should only apply
to versionup to and including 1.5, you could write the following directive:

```python
   license("MIT", when="@:1.5")
```

LydDeb committed:
[fix] py-werkzeug: add constraint in python dependence (#40590)

py-werkzeug@:0.12 does not work with python@3.10:

Test with py-werkzeug 0.12.2 and python 3.10:
```
$ python3.10 -c 'import werkzeug'
py-werkzeug-0.12.2/lib/python3.11/site-packages/werkzeug/datastructures.py", line 16, in <module>
from collections import Container, Iterable, MutableSet
ImportError: cannot import name 'Container' from 'collections'
```

Test with py-werkzeug 0.12.2 and python 3.9:
```
python3.9 -c "from collections import Container"
<string>:1: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated since Python 3.3, and in 3.10 it will stop working
```

haampie committed:
Fix dev-build keep_stage behavior (#40576)

`spack dev-build` would incorrectly set `keep_stage=True` for the
entire DAG, including for non-dev specs, even though the dev specs
have a DIYStage which never deletes sources.

haampie committed:
llvm: fix ncurses+termlib linking in lldb (#40594)

haampie committed:
AutotoolsPackage / MakefilePackage: add gmake build dependency (#40380)

haampie committed:
unparse: also support generic type aliases (#40328)

annop-w committed:
acfl: add version 23.10 (#40510)

michaelkuhn committed:
julia: Fix build for @1.9 (#39045)

julia@1.9 tries to download ittapi, which requires cmake. Disable it
explicitly.

boomanaiden154 committed:
busybox: respect compiler choice (#39239)

boomanaiden154 committed:
bioawk: respect compiler choice (#39241)

boomanaiden154 committed:
connect-proxy: respect compiler choice (#39243)

boomanaiden154 committed:
byte-unixbench: respect compiler choice (#39242)

Co-authored-by: Harmen Stoppels <harmenstoppels@gmail.com>

tldahlgren committed:
Stand-alone test feature deprecation postponed to v0.22 (#40600)

alalazo committed:
ASP-based solver: single Spec instance per dag hash  (#39590)

Reused specs used to be referenced directly into the built spec.

This might cause issues like in issue 39570 where two objects in
memory represent the same node, because two reused specs were
loaded from different sources but referred to the same spec
by DAG hash.

The issue is solved by copying concrete specs to a dictionary keyed
by dag hash.

haampie committed:
ci: remove incorrect compilers.yaml (#40610)

balay committed:
petsc: add variant +sycl (#40562)

* petsc: add variant +sycl

* petsc: add in gmake as dependency - so that consistent make gets used between petsc and slepc builds [that can have different env for each of the builds]

scottwittenburg committed:
gitlab ci: Rework how mirrors are configured (#39939)

Improve how mirrors are used in gitlab ci, where we have until now thought
of them as only a string.

By configuring ci mirrors ahead of time using the proposed mirror templates,
and by taking advantage of the expressiveness that spack now has for mirrors,
this PR will allow us to easily switch the protocol/url we use for fetching
binary dependencies.

This change also deprecates some gitlab functionality and marks it for
removal in Spack 0.23:

    - arguments to "spack ci generate":
        * --buildcache-destination
        * --copy-to
    - gitlab configuration options:
        * enable-artifacts-buildcache
        * temporary-storage-url-prefix

haampie committed:
libvorbis: drop -force_cpusubtype_ALL flag (#40616)

This flag was only relevant when targeting powerpc from apple-clang,
which we don't do. The flag is removed from apple-clang@15. Let's drop
it unconditionally.

vsoch committed:
Automated deployment to update package flux-core 2023-10-19 (#40605)

Co-authored-by: github-actions <github-actions@users.noreply.github.com>

haampie committed:
spack checksum: restore ability to select top n (#40531)

The ability to select the top N versions got removed in the checksum overhaul,
cause initially numbers were used for commands.

Now that we settled on characters for commands, let's make numbers pick the top
N again.

haampie committed:
Improve setup build / run / test environment (#35737)

This adds a `SetupContext` class which is responsible for setting
package.py module globals, and computing the changes to environment
variables for the build, test or run context.

The class uses `effective_deptypes` which takes a list of specs (e.g. single
item of a spec to build, or a list of environment roots) and a context
(build, run, test), and outputs a flat list of specs that affect the
environment together with a flag in what way they do so. This list is
topologically ordered from root to leaf, so that one can be assured that
dependents override variables set by dependencies, not the other way
around.

This is used to replace the logic in `modifications_from_dependencies`,
which has several issues: missing calls to `setup_run_environment`, and
the order in which operations are applied.

Further, it should improve performance a bit in certain cases, since
`effective_deptypes` run in O(v + e) time, whereas `spack env activate`
currently can take up to O(v^2 + e) time due to loops over roots. Each
edge in the DAG is visited once by calling `effective_deptypes` with
`env.concrete_roots()`.

By marking and propagating flags through the DAG, this commit also fixes
a bug where Spack wouldn't call `setup_run_environment` for runtime
dependencies of link dependencies. And this PR ensures that Spack
correctly sets up the runtime environment of direct build dependencies.

Regarding test dependencies: in a build context they are are build-time
test deps, whereas in a test context they are install-time test deps.
Since there are no means to distinguish the build/install type test deps,
they're both.

Further changes:

- all `package.py` module globals are guaranteed to be set before any of the
  `setup_(dependent)_(run|build)_env` functions is called
- traversal order during setup: first the group of externals, then the group
  of non-externals, with specs in each group traversed topological (dependencies
  are setup before dependents)
- modules: only ever call `setup_dependent_run_environment` of *direct* link/run
   type deps
- the marker in `set_module_variables_for_package` is dropped, since we should
  call the method once per spec. This allows us to set only a cheap subset of
  globals on the module: for example it's not necessary to compute the expensive
  `cmake_args` and w/e if the spec under consideration is not the root node to be
  built.
- `spack load`'s `--only` is deprecated (it has no effect now), and `spack load x`
  now means: do everything that's required for `x` to work at runtime, which
  requires runtime deps to be setup -- just like `spack env activate`.
- `spack load` no longer loads build deps (of build deps) ...
- `spack env activate` on partially installed or broken environments: this is all
  or nothing now. If some spec errors during setup of its runtime env, you'll only
  get the unconditional variables + a warning that says the runtime changes for
  specs couldn't be applied.
- Remove traversal in upward direction from `setup_dependent_*` in packages.
  Upward traversal may iterate to specs that aren't children of the roots
  (e.g. zlib / python have hundreds of dependents, only a small fraction is
  reachable from the roots. Packages should only modify the direct dependent
  they receive as an argument)

balos1 committed:
add nvechip to sundials components when mfem+rocm (#40512)

adamjstewart committed:
py-rasterio: add v1.3.9 (#40621)

dependabot[bot] committed:
build(deps): bump urllib3 from 2.0.6 to 2.0.7 in /lib/spack/docs (#40583)

dependabot[bot] committed:
build(deps): bump mypy from 1.6.0 to 1.6.1 in /.github/workflows/style (#40602)

dependabot[bot] committed:
build(deps): bump mypy from 1.6.0 to 1.6.1 in /lib/spack/docs (#40603)

dependabot[bot] committed:
build(deps): bump actions/checkout from 4.1.0 to 4.1.1 (#40584)

LydDeb committed:
perl: change permissions  in order to apply patch on version 5.38.0 (#40609)

Co-authored-by: Harmen Stoppels <harmenstoppels@gmail.com>

vicentebolea committed:
vtk-m: update to latest release (#40624)

wspear committed:
TAU: Respect ~fortran for +mpi (#40617)

adamjstewart committed:
Drop support for external PythonX.Y (#40628)

On some systems, multiple pythonx.y are placed in the same prefix as
pythonx (where only one of them is associated with that pythonx).
Spack external detection for Python was willing to register all of
these as external versions. Moreover, the `package.py` for Python
was able to distinguish these.

This can cause an issue for some build systems, which will just look
for python3 for example, so if that python3 is actually python3.6,
and the build system needs 3.7 (which spack may have found in the
same prefix, and offered as a suitable external), it will fail when
invoking python3.

To avoid that issue, we simply avoid treating pythonx.y as external
candidates. In the above case, Spack would only detect a Python 3.6
external, and the build would be forced to use a Spack-built Python
3.7 (which we consider a good thing).

haampie committed:
schema/compilers.py: fix validation of 2+ entries (#40627)

Fix the following syntax which validates only the first array entry:

```python
"compilers": {
    "type": "array",
    "items": [
        {
            "type": ...
        }
    ]
}
```

to

```python
"compilers": {
    "type": "array",
    "items": {
        "type": ...
    }
}
```

which validates the entire array.

Oops...

snehring committed:
Add package py-macs3 and dependencies (#40498)

* py-cykhash: adding new package py-cykhash

* py-hmmlearn: adding new package py-hmmlearn

* py-macs3: adding new package py-macs3

* py-macs3: adding python version restriction and other changes.

manuelakuhn committed:
py-bidscoin: add v4.1.1 and py-argparse-manpage: add new package (#40414)

* py-bidscoin: add 4.1.1

* Fix style

* Fix restrictions for dependencies

manuelakuhn committed:
py-dcm2bids: add v3.1.0 (#40447)

* py-dcm2bids: add 3.1.0

* Fix python restriction

manuelakuhn committed:
py-statsmodels: add 0.14.0 (#39156)

* py-statsmodels: add 0.14.0

* Fix style

* Update var/spack/repos/builtin/packages/py-statsmodels/package.py

Co-authored-by: Adam J. Stewart <ajstewart426@gmail.com>

* Update var/spack/repos/builtin/packages/py-statsmodels/package.py

Co-authored-by: Adam J. Stewart <ajstewart426@gmail.com>

* Remove python limits

* Remove comment

---------

Co-authored-by: Adam J. Stewart <ajstewart426@gmail.com>

LydDeb committed:
[add] py-css-parser: new package (#40550)

Co-authored-by: LydDeb <lyderic.debusschere@eolen.com>

LydDeb committed:
[add] py-cssutils: new package (#40551)

Co-authored-by: LydDeb <lyderic.debusschere@eolen.com>

LydDeb committed:
[add] py-fraction: new package (#40554)

Co-authored-by: LydDeb <lyderic.debusschere@eolen.com>

Bidibulke committed:
add recipes for sphinx-book-theme and its dependencies if unknown (#40312)

* add recipes for sphinx-book-theme and its dependencies if unknown

* fix version and mission https

* fix based on reviewers remarks

adamjstewart committed:
TensorFlow/Keras/TensorBoard: add v2.14.0 (#40297)

Co-authored-by: adamjstewart <adamjstewart@users.noreply.github.com>

alalazo committed:
ASP-based solver: minimize weights over edges (#40632)

With the introduction of multiple build dependencies from the same package in the DAG, we need to minimize a few weights accounting for edges rather than nodes. If we don't do that we might have multiple "optimal" solutions that differ only in how the same nodes are connected together. This commit ensures optimal versions are picked per parent in case of multiple choices for a dependency.

LydDeb committed:
py-corner: new package (#40546)

* [add] py-corner: new package

* py-corner: remove py-wheel dependence with respect to reviewing commentary

---------

Co-authored-by: LydDeb <lyderic.debusschere@eolen.com>

haampie committed:
gromacs +cp2k: build in CI (#40494)

* gromacs +cp2k: build in CI

* libxsmm: x86 only

* attempt to fix dbcsr + new mpich

* use c11 standard

* gromacs: does not depend on dbcsr

* cp2k: build with cmake in CI, s.t. dbcsr is a separate package

* cp2k: cmake patches for config files and C/C++ std

* cp2k: remove unnecessary constraints due to patch

haampie committed:
concretize separately: show concretization time per spec as they concretize when verbose (#40634)

garth-wells committed:
py-fenics-ffcx: update to v0.7 (#40569)

al42and committed:
gromacs: add 2022.6, 2023.2, 2023.3 versions (#38906)

* gromacs: add 2022.6, 2023.2 versions
* gromacs: add version 2023.3

LydDeb committed:
[add] py-dict2css: new package (#40552)

Co-authored-by: LydDeb <lyderic.debusschere@eolen.com>

scottwittenburg committed:
py-kombu: pick older version of py-setuptools (#40642)

haampie committed:
py-kombu: fix setuptools bound (#40646)

aumuell committed:
intel-tbb: patch patch for Apple's patch (#40640)

While e.g. GNU patch 2.7.6 (as provided by homebrew) would apply the previous
version of this patch without problems, Apple's patch 2.0-12u11-Apple fails
to find out which file to patch.

Adding two lines to the patch fixes that. Renamed the patch in order to
not require a `spack clean -m`.

tldahlgren committed:
Docs: Add version range example to conditional dependencies (#40630)

* Docs: Add version range example to conditional dependencies

* Add when context manager example

jmcarcell committed:
py-kiwisolver: add a new version (#40653)

Co-authored-by: jmcarcell <jmcarcell@users.noreply.github.com>

haampie committed:
zlib-ng: add v2.1.4 (#40647)

jmcarcell committed:
glib: add patch with a fix for PTRACE_0_EXITKILL (#40655)

Co-authored-by: jmcarcell <jmcarcell@users.noreply.github.com>

wrwilliams committed:
Add Score-P 8.3 and dependencies (#40478)

Includes Score-P 8.3 and Cubew/cubelib 4.8.2.

michaelkuhn committed:
libxml2: fix GitLab patch (#40658)

GitLab's .patch URLs do not provide stable/full hashes, while .diff URLs
do. See #40656 for more information.

michaelkuhn committed:
vtk: fix GitLab patch (#40659)

GitLab's .patch URLs do not provide stable/full hashes, while .diff URLs
do. See #40656 for more information.

michaelkuhn committed:
garfieldpp: fix GitLab patch (#40660)

GitLab's .patch URLs do not provide stable/full hashes, while .diff URLs
do. See #40656 for more information.

michaelkuhn committed:
gobject-introspection: fix GitLab patch (#40661)

GitLab's .patch URLs do not provide stable/full hashes, while .diff URLs
do. See #40656 for more information.

michaelkuhn committed:
knem: fix GitLab patch (#40662)

GitLab's .patch URLs do not provide stable/full hashes, while .diff URLs
do. See #40656 for more information.

michaelkuhn committed:
libtheora: fix GitLab patch (#40657)

GitLab's .patch URLs do not provide stable/full hashes, while .diff URLs
do. See #40656 for more information.

tldahlgren committed:
Docs: Update spec variant checks plus python quotes and string formatting (#40643)

haampie committed:
py-cython: new version, python 3.11 upperbound (#40343)

haampie committed:
concretizer verbose: show progress in % too (#40654)

boomanaiden154 committed:
3proxy: respect compiler choice (#39240)

haampie committed:
nghttp2: add v1.57.0 (#40652)

jmcarcell committed:
geant4: add patch for when using the system expat library (#40650)

Co-authored-by: jmcarcell <jmcarcell@users.noreply.github.com>

cessenat committed:
ngspice: new version 41 and option osdi (#40664)

michaelkuhn committed:
audit: add check for GitLab patches (#40656)

GitLab's .patch URLs only provide abbreviated hashes, while .diff URLs
provide full hashes. There does not seem to be a parameter to force
.patch URLs to also return full hashes, so we should make sure to use
the .diff ones.

adamjstewart committed:
py-scikit-learn: add v1.3.2 (#40672)

jgalarowicz committed:
Update survey package file for survey version 9 changes. (#40619)

* Update survey package file for survey version 9 changes.
* Fix single quote - make double.
* Small change to trigger spack tests

mtaillefumier committed:
Add rccl and nccl variants to cp2k and cosma (#40451)

vicentebolea committed:
Adios2: add kokkos variant (#40623)

* adios2: update variants and dependencies

* adios2: add kokkos rocm|cuda|sycl variant

* e4s oneapi ci stack: add adios2 +sycl

* e4s ci stack: add adios2 +rocm

* [@spackbot] updating style on behalf of vicentebolea

* Apply suggestions from code review

* adios2: fixed cuda variant

* update ecp-data-vis-sdk

* Update share/spack/gitlab/cloud_pipelines/stacks/e4s-power/spack.yaml

---------

Co-authored-by: eugeneswalker <eugenesunsetwalker@gmail.com>
Co-authored-by: vicentebolea <vicentebolea@users.noreply.github.com>

namachan10777 committed:
fix installation error of bear (#40637)

Co-authored-by: Tom Scogland <scogland1@llnl.gov>

annop-w committed:
armpl-gcc: add version 23.10 and macOS support (#40511)

eugeneswalker committed:
exago: fix v1.5.1 tag; only allow python up to 3.10 for for @:1.5 (#40676)

* exago: fix v1.5.1 tag; only allow python up to 3.10 for for @:1.5 due to pybind error with py 3.11

* hiop@:1.0 +cuda: constrain to cuda@:11.9

eugeneswalker committed:
hiop +cuda: fix issue 40678 (#40688)

albestro committed:
neovim: conflict for libluv problem on macOS + add newer versions of neovim and libluv (#40690)

* add conflict with libluv version >=1.44 just on macOS
* minor change
* add libluv versions
* neovim: add newer releases

fbarbari committed:
Added new benchmark version up to 1.8.3 (#40689)

amd-toolchain-support committed:
openmpi: fix pmi@4.2.3: compat (#40686)

renjithravindrankannath committed:
Updating rvs binary path. (#40604)

* Updating rvs binary path
* Updating spec check as per the recommendation

AlexanderRichert-NOAA committed:
Add ufs-utils@1.11.0 (#40695)

* Add ufs-utils@1.11.0
* Update package.py

johnwparent committed:
Windows: search PATH for patch utility (#40513)

Previously, we only searched for `patch` inside of whatever Git
installation was available because the most common installation of Git
available on Windows had `patch`. That's not true for all possible
installations of Git though, so this updates the search to also check
PATH.

mtaillefumier committed:
[cp2k] Use fftw3 MKL by default when cp2k is compiled with mkl (#40671)

RMeli committed:
Add dlaf variant to cp2k (#40702)

adamjstewart committed:
py-lightning: py-torch~distributed is broken again (#40696)

haampie committed:
ci: darwin aarch64 use apple-clang-15 tag (#40706)

haampie committed:
ci: don't put compilers in config (#40700)

* ci: don't register detectable compilers

Cause they go out of sync...

* remove intel compiler, it can be detected too

* Do not run spack compiler find since compilers are registered in concretize job already

* trilinos: work around +stokhos +cuda +superlu-dist bug due to EMPTY macro

dependabot[bot] committed:
build(deps): bump pytest from 7.4.2 to 7.4.3 in /lib/spack/docs (#40697)

adamjstewart committed:
PyTorch: patch breakpad dependency (#40648)

fbarbari committed:
Added Highway versions up to 1.0.7 (#40691)

dominichofer committed:
cuda: add NVHPC_CUDA_HOME. (#40507)

* [cuda] Add NVHPC_CUDA_HOME.

* Add CUDA_HOME and NVHC_CUDA_HOME to cuda's dependent build env.

---------

Co-authored-by: Dominic Hofer <dominic.hofer@meteoswiss.ch>

afzpatel committed:
initial commit to fix mivisionx build for 5.6 (#40579)

victoria-cherkas committed:
fdb: add releases v5.11.23 and v5.11.17 (#40571)

mathstuf committed:
Paraview 5.12 prep (#40527)

* paraview: rebase the adios2 patch for 5.12-to-be

* paraview: disable fastfloat and token for 5.12-to-be

* paraview: require older protobuf for 5.12 as well

* paraview: require C++11-supporting protobuf for `master` too

skosukhin commented:
As I have already mentioned, it's hard for me to find the motivation to maintain this package while having a 20-month pending PR for it (see #29299 and #36186).

haampie committed:
git versions: fix commit shas [automated] (#40703)

albestro committed:
libluv: require CMake 3 and CMP0042 (#40716)

xdelaruelle committed:
modules: hide implicit modulefiles (#36619)

Renames exclude_implicits to hide_implicits

When hide_implicits option is enabled, generate modulefile of
implicitly installed software and hide them. Even if implicit, those
modulefiles may be referred as dependency in other modulefiles thus they
should be generated to make module properly load dependent module.

A new hidden property is added to BaseConfiguration class.

To hide modulefiles, modulercs are generated along modulefiles. Such rc
files contain specific module command to indicate a module should be
hidden (for instance when using "module avail").

A modulerc property is added to TclFileLayout and LmodFileLayout classes
to get fully qualified path name of the modulerc associated to a given
modulefile.

Modulerc files will be located in each module directory, next to the
version modulefiles. This scheme is supported by both module tool
implementations.

modulerc_header and hide_cmd_format attributes are added to
TclModulefileWriter and LmodModulefileWriter. They help to know how to
generate a modulerc file with hidden commands for each module tool.

Tcl modulerc file requires an header. As we use a command introduced on
Modules 4.7 (module-hide --hidden-loaded), a version requirement is
added to header string.

For lmod, modules that open up a hierarchy are never hidden, even if
they are implicitly installed.

Modulerc is created, updated or removed when associated modulefile is
written or removed. If an implicit modulefile becomes explicit, hidden
command in modulerc for this modulefile is removed. If modulerc becomes
empty, this file is removed. Modulerc file is not rewritten when no
content change is detected.

Co-authored-by: Harmen Stoppels <me@harmenstoppels.nl>

haampie committed:
spack checksum: show long flags in usage output (#40407)

adamjstewart committed:
PythonPackage: nested config_settings (#40693)

* PythonPackage: nested config_settings

* flake8

xdelaruelle committed:
modules: no --delim option if separator is colon character (#39010)

Update Tcl modulefile template to simplify generated `append-path`,
`prepend-path` and `remove-path` commands and improve their readability.

If path element delimiter is colon character, do not set the `--delim`
option as it is the default delimiter value.

aurianer committed:
Add conflict between cxxstd > 17 and cuda < 12 in pika (#40717)

* Add conflict with C++ standard > 17 and cuda < 12

* Removing map_cxxstd since boost supports C++20 flag

haampie committed:
spack checksum pkg@1.2, use as version filter (#39694)

* spack checksum pkg@1.2, use as version filter

Currently pkg@1.2 splits on @ and looks for 1.2 specifically, with this
PR pkg@1.2 is a filter so any matching 1.2, 1.2.1, ..., 1.2.10 version
is displayed.

* fix tests

* fix style

ryandanehy committed:
Update spack package for exago@1.6.0 release (#40614)

* Update spack package for exago:1.6.0

* update style

* Weird spack style env bug fixed

* Update spack package for exago:1.6.0

* update style

* Weird spack style env bug fixed

* changes to allow release 1.6.0

* fix depends, and versioning

* rm cmake variable

* add s

* style fix

---------

Co-authored-by: Ryan Danehy <dane678@deception04.pnl.gov>
Co-authored-by: Ryan Danehy <dane678@deception03.pnl.gov>
Co-authored-by: ryan.danehy@pnnl.gov <dane678@we45149.home>

alecbcs committed:
unmaintained a* packages: update to use f-strings (#40467)

tobbez committed:
curl: Fix librtmp variant (#40713)

* rtmpdump: New package

* curl: Fix librtmp variant

Add the previously missing dependency required for rtmp support.

The variant has been broken since its addition in PR #25166.

Fixes one of the two issues reported in #26887.

adamjstewart committed:
PythonPackage: allow archive_files to be overridden (#40694)

masterleinad committed:
dataTransferKit: add v3.1.1, v3.1.0 (#40556)

* Update DataTransferKit for 3.1.1 release

* Require Trilinos-14 for 3.1.0 and higher

snehring committed:
ldak: add v5.2 & add maintainer (#40710)

* ldak: update to 5.2, add maintainer

* ldak: use compiler.openmp_flag

zzzoom committed:
itk: misc fixes (#39832)

* itk: patch missing include for newer compilers

* itk: The package doesn't use MPI

* itk: package requires the high-level hdf5 api

* itk: patch url with ?full_index=1

* itk: point to 4041 commit in master

* itk: don't constrain hdf5 with ~mpi

balay committed:
plasma: add version 23.8.2 (#40728)

alecbcs committed:
akantu: use f-strings (#40466)

Co-authored-by: Nicolas Richart <nrichart@users.noreply.github.com>

haampie committed:
gromacs: fix version branch in intel fftw (#40489)

dmt4 committed:
Fixes and options for package spglib (#40684)

* Fix cmake_args for spglib v2.1.0+

* Add option to build fortran interface in package spglib

* fix style as sugested by ci/prechecks/style

* Enable fortran variant from v1.16.4 as suggested

Co-authored-by: Rocco Meli <r.meli@bluemail.ch>

---------

Co-authored-by: Rocco Meli <r.meli@bluemail.ch>

haampie committed:
ci: spack compiler find should list extra config scopes (#40727)

otherwise it detected pre-configured compilers in an potentially different way.

haampie committed:
gromacs: default to external blas & lapack (#40490)

* gromacs: default to external blas & lapack

* drop vendored lapack/blas altogether

balay committed:
strumpack: add version 7.2.0 (#40732)

iamashwin99 committed:
octopus: split netcdf-c and netcdf-fortran dependency (#40685)

victoria-cherkas committed:
modify aec variant so as not to change the current behaviour

haampie committed:
OCI buildcache (#38358)

Credits to @ChristianKniep for advocating the idea of OCI image layers
being identical to spack buildcache tarballs.

With this you can configure an OCI registry as a buildcache:

```console 
$ spack mirror add my_registry oci://user/image # Dockerhub

$ spack mirror add my_registry oci://ghcr.io/haampie/spack-test # GHCR

$ spack mirror set --push --oci-username ... --oci-password ... my_registry  # set login credentials
```

which should result in this config:

```yaml
mirrors:
  my_registry:
    url: oci://ghcr.io/haampie/spack-test
    push:
      access_pair: [<username>, <password>]
```

It can be used like any other registry

```
spack buildcache push my_registry [specs...]
```

It will upload the Spack tarballs in parallel, as well as manifest + config
files s.t. the binaries are compatible with `docker pull` or `skopeo copy`.

In fact, a base image can be added to get a _runnable_ image:

```console
$ spack buildcache push --base-image ubuntu:23.04 my_registry python
Pushed ... as [image]:python-3.11.2-65txfcpqbmpawclvtasuog4yzmxwaoia.spack

$ docker run --rm -it [image]:python-3.11.2-65txfcpqbmpawclvtasuog4yzmxwaoia.spack
```

which should really be a game changer for sharing binaries.

Further, all content-addressable blobs that are downloaded and verified
will be cached in Spack's download cache. This should make repeated
`push` commands faster, as well as `push` followed by a separate
`update-index` command.

An end to end example of how to use this in Github Actions is here:

**https://github.com/haampie/spack-oci-buildcache-example**


TODO:

- [x] Generate environment modifications in config so PATH is set up
- [x] Enrich config with Spack's `spec` json (this is allowed in the OCI specification)
- [x] When ^ is done, add logic to create an index in say `<image>:index` by fetching all config files (using OCI distribution discovery API)
- [x] Add logic to use object storage in an OCI registry in `spack install`.
- [x] Make the user pick the base image for generated OCI images.
- [x] Update buildcache install logic to deal with absolute paths in tarballs
- [x] Merge with `spack buildcache` command
- [x] Merge #37441 (included here)
- [x] Merge #39077 (included here)
- [x] #39187 + #39285
- [x] #39341
- [x] Not a blocker: #35737 fixes correctness run env for the generated container images

NOTE:

1. `oci://` is unfortunately taken, so it's being abused in this PR to mean "oci type mirror". `skopeo` uses `docker://` which I'd like to avoid, given that classical docker v1 registries are not supported.
2. this is currently `https`-only, given that basic auth is used to login. I _could_ be convinced to allow http, but I'd prefer not to, given that for a `spack buildcache push` command multiple domains can be involved (auth server, source of base image, destination registry). Right now, no urllib http handler is added, so redirects to https and auth servers with http urls will simply result in a hard failure.

CAVEATS:

1. Signing is not implemented in this PR. `gpg --clearsign` is not the nicest solution, since (a) the spec.json is merged into the image config, which must be valid json, and (b) it would be better to sign the manifest (referencing both config/spec file and tarball) using more conventional image signing tools
2. `spack.binary_distribution.push` is not yet implemented for the OCI buildcache, only `spack buildcache push` is. This is because I'd like to always push images + deps to the registry, so that it's `docker pull`-able, whereas in `spack ci` we really wanna push an individual package without its deps to say `pr-xyz`, while its deps reside in some `develop` buildcache.
3. The `push -j ...` flag only works for OCI buildcache, not for others

johnwparent committed:
External finding: update default paths; treat .bat as executable on Windows (#39850)

.bat or .exe files can be considered executable on Windows. This PR
expands the regex for detectable packages to allow for the detection
of packages that vendor .bat wrappers (intel mpi for example).

Additional changes:

* Outside of Windows, when searching for executables `path_hints=None`
  was used to indicate that default path hints should be provided,
  and `[]` was taken to mean that no defaults should be chosen
  (in that case, nothing is searched); behavior on Windows has
  now been updated to match.
* Above logic for handling of `path_hints=[]`  has also been extended
  to library search (for both Linux and Windows).
* All exceptions for external packages were documented as timeout
  errors: this commit adds a distinction for other types of errors
  in warning messages to the user.

eugeneswalker committed:
mgard@2020-10-01 %oneapi@2023: turn of c++11-narrowing via cxxflags (#40743)

eugeneswalker committed:
e4s ci stacks: add exago specs (#40712)

* e4s ci: add exago +cuda, +rocm builds

* exago: rename 5-18-2022-snapshot to snapshot.5-18-2022

* disable exago +rocm for non-external rocm ci install

* note that hiop +rocm fails to find hip libraries when they are spack-installed

giordano committed:
hipsycl: restrict compatibility with llvm for v0.8.0 (#40736)

johnwparent committed:
MSVC: detection from registry (#38500)

Typically MSVC is detected via the VSWhere program. However, this may
not be available, or may be installed in an unpredictable location.
This PR adds an additional approach via Windows Registry queries to
determine VS install location root.

Additionally:

* Construct vs_install_paths after class-definition time (move it to
  variable-access time).
* Skip over keys for which a user does not have read permissions
  when performing searches (previously the presence of these keys
  would have caused an error, regardless of whether they were
  needed).
* Extend helper functionality with option for regex matching on
  registry keys vs. exact string matching.
* Some internal refactoring: remove boolean parameters in some cases
  where the function was always called with the same value
  (e.g. `find_subkey`)

adamjstewart committed:
py-numpy: add v1.26 (#40057)

manuelakuhn committed:
py-comm: add 0.1.4 (#40669)

heerener committed:
py-bluepyemodel: opensourcing with dependencies (#40592)

* py-bluepyemodel: new package with dependencies

* py-morphio: add MPI as dependency to avoid failing builds

* Formatting

* py-bluepyefe: no need to set NEURON_INIT_MPI

* py-morphio: unifurcation branch is ancient history

* py-bluepyopt: only set NEURON_INIT_MPI with +neuron

* py-efel: get rid of old version

* py-morph{-tool,io}: rename develop to master to match branch

* py-bluepyefe: unset PMI_RANK is also neuron-related

* py-bluepyopt: PMI_RANK is also neuron-related

* Implement review remarks

* py-morph-tool, py-neurom: small fixes

* py-morphio: reword dependencies

qwertos committed:
New version of py-langsmith (#40674)

Co-authored-by: Benjamin Meyers <bsmits@rit.edu>

RemiLacroix-IDRIS committed:
NCCL: Add version 2.19.3-1 (#40704)

LydDeb committed:
py-generateds: new package (#40555)

* [add] py-generateds: new package

* py-generateds: Update from review

Co-authored-by: Manuela Kuhn <36827019+manuelakuhn@users.noreply.github.com>

* py-generateds: add versions 2.41.5, 2.42.1, 2.42.2, 2.43.1 and 2.43.2

---------

Co-authored-by: LydDeb <lyderic.debusschere@eolen.com>
Co-authored-by: Manuela Kuhn <36827019+manuelakuhn@users.noreply.github.com>

LydDeb committed:
py-moarchiving: new package (#40558)

* [add] py-moarchiving: new package

* py-moarchiving: update from review: description, variant default value is False, switch when and type

---------

Co-authored-by: LydDeb <lyderic.debusschere@eolen.com>

soumagne committed:
mercury: add v2.3.1 (#40749)

SofiaXu committed:
Add liggght patched for newer compiler (#38685)

* Add liggght patched for newer compiler

Add C++ 17 support
Add Clang and Oneapi support

* Add maintainers

* Fix format in liggghts

* Fix maintainers before versions

Co-authored-by: Alec Scott <alec@bcs.sh>

* Fix style and user to usr

* Update package.py

---------

Co-authored-by: Alec Scott <alec@bcs.sh>

haampie committed:
py-numcodecs: fix broken sse / avx2 variables (#40754)

wdconinc committed:
acts: add v28.1.0:30.3.2 (#40723)

* acts: new version from 28.1.0 to 30.3.1

* acts: new version 30.3.2

* acts: new variant +podio

alalazo committed:
ASP-based solver: avoid cycles in clingo using hidden directive (#40720)

The code should be functonally equivalent to what it was before,
but now to avoid cycles by design we are using a "hidden"
feature of clingo

v-dobrev committed:
mfem: support petsc+rocm with spack-installed rocm (#40768)

Commit:
exago: fix exago missing on PYTHONPATH when `+python` (#40748)

aumuell committed:
pcl: checksum new versions (#39039)

adamjstewart committed:
PyTorch: build with external fp16 (#40760)

adamjstewart committed:
py-pandas: add v2.1.2 (#40734)

v-dobrev committed:
Fix an issue with using the environment variable `MACHTYPE` which is not always defined (#40733)

* Fix an issue reported here:
   https://github.com/spack/spack/pull/36154#issuecomment-1781854894

* [@spackbot] updating style on behalf of v-dobrev

victoria-cherkas committed:
style formatting

wspear committed:
RAJA: add "plugins" variant (#40750)

victoria-cherkas committed:
formatting

victoria-cherkas commented:
Sorry i tried to rebase on develop because the style formatting was working locally but not on CI checks and I've messed up. Closing and will reopen a new PR on new branch.
